### PR TITLE
Add watch namespaces option to the controller

### LINKED ISSUE #620: Support include/select namespaces in controller
## Bug Report

I see lot of activities in controller logs. Why is controller trying to access ingress namespace or any other namespaces?

I see ignoreNamespaces option in helm chart. is it possible to have includeNamespaces/selectNamespaces so that I can define which namespaces maesh should pick up?

### Output of controller log: (_1.2.0_)

```
maesh-controller-5cd7d56465-2cfv2 maesh-controller 2020-05-31T05:47:07.224836129Z time="2020-05-31T05:47:07Z" level=error msg="Could not update mesh service: unable to update service \"maesh-nginx-ingress-controller-default-backend-6d61657368-ingress\": unable to get shadow service \"maesh-nginx-ingress-controller-default-backend-6d61657368-ingress\": service \"maesh-nginx-ingress-controller-default-backend-6d61657368-ingress\" not found"
maesh-controller-5cd7d56465-2cfv2 maesh-controller 2020-05-31T05:47:07.257586633Z time="2020-05-31T05:47:07Z" level=error msg="Could not update mesh service: unable to update service \"maesh-prometheus-operator-kube-state-metrics-6d61657368-monitoring\": unable to get shadow service \"maesh-prometheus-operator-kube-state-metrics-6d61657368-monitoring\": service \"maesh-prometheus-operator-kube-state-metrics-6d61657368-monitoring\" not found"
maesh-controller-5cd7d56465-2cfv2 maesh-controller 2020-05-31T05:47:09.545985318Z time="2020-05-31T05:47:09Z" level=error msg="Could not update mesh service: unable to update service \"maesh-prometheus-operator-prometheus-node-exporter-6d61657368-monitoring\": unable to get shadow service \"maesh-prometheus-operator-prometheus-node-exporter-6d61657368-monitoring\": service \"maesh-prometheus-operator-prometheus-node-exporter-6d61657368-monitoring\" not found"
maesh-controller-5cd7d56465-2cfv2 maesh-controller 2020-05-31T05:52:07.33331752Z time="2020-05-31T05:52:07Z" level=error msg="Could not update mesh service: unable to update service \"maesh-prometheus-operator-kube-state-metrics-6d61657368-monitoring\": unable to get shadow service \"maesh-prometheus-operator-kube-state-metrics-6d61657368-monitoring\": service \"maesh-prometheus-operator-kube-state-metrics-6d61657368-monitoring\" not found"
maesh-controller-5cd7d56465-2cfv2 maesh-controller 2020-05-31T05:52:09.546040152Z time="2020-05-31T05:52:09Z" level=error msg="Could not update mesh service: unable to update service \"maesh-prometheus-operator-prometheus-node-exporter-6d61657368-monitoring\": unable to get shadow service \"maesh-prometheus-operator-prometheus-node-exporter-6d61657368-monitoring\": service \"maesh-prometheus-operator-prometheus-node-exporter-6d61657368-monitoring\" not found"
maesh-controller-5cd7d56465-2cfv2 maesh-controller 2020-05-31T05:52:10.746485066Z time="2020-05-31T05:52:10Z" level=error msg="Could not update mesh service: unable to update service \"maesh-nginx-ingress-controller-default-backend-6d61657368-ingress\": unable to get shadow service \"maesh-nginx-ingress-controller-default-backend-6d61657368-ingress\": service \"maesh-nginx-ingress-controller-default-backend-6d61657368-ingress\" not found"
maesh-controller-5cd7d56465-2cfv2 maesh-controller 2020-05-31T05:57:07.239403518Z time="2020-05-31T05:57:07Z" level=error msg="Could not update mesh service: unable to update service \"maesh-prometheus-operator-kube-state-metrics-6d61657368-monitoring\": unable to get shadow service \"maesh-prometheus-operator-kube-state-metrics-6d61657368-monitoring\": service \"maesh-prometheus-operator-kube-state-metrics-6d61657368-monitoring\" not found"
maesh-controller-5cd7d56465-2cfv2 maesh-controller 2020-05-31T05:57:08.746571376Z time="2020-05-31T05:57:08Z" level=error msg="Could not update mesh service: unable to update service \"maesh-prometheus-operator-prometheus-node-exporter-6d61657368-monitoring\": unable to get shadow service \"maesh-prometheus-operator-prometheus-node-exporter-6d61657368-monitoring\": service \"maesh-prometheus-operator-prometheus-node-exporter-6d61657368-monitoring\" not found"
maesh-controller-5cd7d56465-2cfv2 maesh-controller 2020-05-31T05:57:09.94668693Z time="2020-05-31T05:57:09Z" level=error msg="Could not update mesh service: unable to update service \"maesh-nginx-ingress-controller-default-backend-6d61657368-ingress\": unable to get shadow service \"maesh-nginx-ingress-controller-default-backend-6d61657368-ingress\": service \"maesh-nginx-ingress-controller-default-backend-6d61657368-ingress\" not found"
```

### What is your environment & configuration (arguments, provider, platform, ...)?

EKS 
k8s 1.16
maesh 1.2.0

### PR DESCRIPTION
**Pull request by jspdown**
## What does this PR do?

This PR:
- Adds the `watchNamespaces` option to the controller. This allow to watch a limited set of resources.
- Adds the `controller.watchNamespaces` option to the helm chart.
- Refactor the IgnoreWrapper and replace it with a ResourceFilter.
- Refactor the PortMapper so we can stop exposing the ServicePort struct.

Fixes #620

### How to test it

* `make test`
* Install Maesh with the `watchNamespaces` option set. And make sure the controller listen only for events coming from the namespace specified.

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -100,16 +101,13 @@ func NewMeshController(clients k8s.Client, cfg Config, store SharedStore, logger
...
-		c.ignoredResources.AddIgnoredNamespace(ns)
-	}
-
-	c.ignoredResources.AddIgnoredService("kubernetes", metav1.NamespaceDefault)
-	c.ignoredResources.AddIgnoredNamespace(metav1.NamespaceSystem)
-	c.ignoredResources.AddIgnoredApps("maesh", "jaeger")
+	// Initialize the ignored and watched resources.
+	c.resourceFilter = k8s.NewResourceFilter(
+		k8s.WatchNamespaces(cfg.WatchNamespaces...),
+		k8s.IgnoreNamespaces(cfg.IgnoreNamespaces...),
```
[the remark this answers is not in the document]
Comment (jspdown) [reply pkg/controller/controller.go:107]: Sadly, it doesn't work in Go:
example: https://play.golang.org/p/Js9_aZ4BnKr
doc: https://golang.org/ref/spec#Passing_arguments_to_..._parameters

We could eventually write it like:
```
k8s.IgnoreNamespaces(append(metav1.NamespaceSystem, cfg.IgnoreNamespaces...)...)
```

But I find this notation a but difficult to read.

[the remark this answers is not in the document]
Comment (kevinpollet) [reply pkg/controller/controller.go:107]: I thought it works =)
I like your approach, sorry for that!

File: pkg/k8s/filter_test.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestNewResourceFilter(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (jspdown) [reply pkg/k8s/filter_test.go:11]: I'm not sure we want to do this. The function we are testing is `NewResourceFilter`.

File: pkg/provider/provider_test.go [hunk 3]
```diff
@@ -5,38 +5,43 @@ import (
...
-	mk8s "github.com/containous/maesh/pkg/k8s"
 	"github.com/containous/maesh/pkg/topology"
 	"github.com/containous/traefik/v2/pkg/config/dynamic"
 	"github.com/sirupsen/logrus"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
-type stateTableMock func(svcPort mk8s.ServicePort) (int32, bool)
+type stateTableMock func(ns, name string, port int32) (int32, bool)
```
Comment (kevinpollet) [inline pkg/provider/provider_test.go:15]: Can you rename all `ns` param in this file?

```suggestion
type stateTableMock func(namespace, name string, port int32) (int32, bool)
```

File: pkg/controller/service_test.go [hunk 4]
```diff
@@ -20,33 +19,33 @@ import (
 )
 
 type portMapperMock struct {
-	findFunc   func(svc k8s.ServicePort) (int32, bool)
-	addFunc    func(svc *k8s.ServicePort) (int32, error)
-	removeFunc func(svc k8s.ServicePort) (int32, error)
+	findFunc   func(ns, name string, port int32) (int32, bool)
```
Comment (kevinpollet) [inline pkg/controller/service_test.go:22]: Can you rename all ns param in this file?

```suggestion
	findFunc   func(namespace, name string, port int32) (int32, bool)
```

**Pull request merged**